app/over_react_redux/todo_client/lib/src/store.dart

```diff
@@ -0,0 +1,190 @@
+import 'package:json_annotation/json_annotation.dart';
+import 'package:over_react/over_react_redux.dart';
+import 'package:redux/redux.dart';
+import 'package:redux_dev_tools/redux_dev_tools.dart';
+
+import 'package:todo_client/src/models/todo.dart';
+import 'package:todo_client/src/models/user.dart';
+import 'package:todo_client/src/actions.dart';
+import 'package:todo_client/src/local_storage.dart';
+
+part 'store.g.dart';
+
+AppState _initializeState() {
+  AppState initialState;
+  if (!TodoAppLocalStorage.isInitialized) {
+    // First load - give the user some data to work with, and set up our default / empty states.
+    initialState = AppState.fromJson(defaultAppState);
+    localTodoAppStorage = TodoAppLocalStorage(initialState);
+  } else {
+    localTodoAppStorage ??= TodoAppLocalStorage();
+    initialState = AppState.fromJson(localTodoAppStorage.currentStateJson);
+  }
+
+  return initialState;
+}
+
+var store = DevToolsStore<AppState>(
+  appStateReducer,
+  initialState: _initializeState(),
+  middleware: [overReactReduxDevToolsMiddleware],
+);
+
+@JsonSerializable(explicitToJson: true)
+class AppState {
+  String name;
+  List<Todo> todos;
+  List<User> users;
+  List<String> selectedTodoIds;
+  List<String> editableTodoIds;
+  List<String> highlightedTodoIds;
+  List<String> selectedUserIds;
+  List<String> editableUserIds;
+  List<String> highlightedUserIds;
+
+  AppState(this.name, {
+    this.todos,
+    this.users,
+    this.selectedTodoIds,
+    this.editableTodoIds,
+    this.highlightedTodoIds,
+    this.selectedUserIds,
+    this.editableUserIds,
+    this.highlightedUserIds,
+  }) {
+    assert(name != null);
+    localTodoAppStorage?.updateCurrentState(this);
+  }
+
+  factory AppState.fromJson(Map<String, dynamic> json) => _$AppStateFromJson(json);
+  Map<String, dynamic> toJson() => _$AppStateToJson(this);
+}
+
+AppState appStateReducer(AppState state, dynamic action) {
+  if (action is SaveLocalStorageStateAsAction) {
+    Map previousValue;
+    if (action.value.previousName != null) {
+      previousValue = localTodoAppStorage.remove(action.value.previousName);
+    } else {
+      previousValue = localTodoAppStorage.currentStateJson;
+    }
+
+    localTodoAppStorage[action.value.name] = (AppState.fromJson(previousValue)..name = action.value.name).toJson();
+  }
+
+  if (action is LoadStateFromLocalStorageAction) {
+    return AppState.fromJson(localTodoAppStorage[action.value]);
+  }
+
+  return AppState(localTodoAppStorage.currentStateJson['name'],
```

Comment:
This seems like it could be an impure reducer; am I correct in this? Could we achieve this some other way (via its own reducer or a middleware or something?